instrumentation/jaxws/jaxws-2.0-metro-2.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/metro/MetroJaxWsTracer.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.metro;
+
+import static io.opentelemetry.api.trace.SpanKind.INTERNAL;
+
+import com.sun.xml.ws.api.message.Packet;
+import com.sun.xml.ws.api.server.WSEndpoint;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.instrumentation.api.servlet.ServletContextPath;
+import io.opentelemetry.instrumentation.api.tracer.BaseTracer;
+import io.opentelemetry.javaagent.instrumentation.metro.TracingPropertySet.ThrowableHolder;
+import javax.servlet.http.HttpServletRequest;
+import javax.xml.ws.handler.MessageContext;
+
+public class MetroJaxWsTracer extends BaseTracer {
+  private static final MetroJaxWsTracer TRACER = new MetroJaxWsTracer();
+
+  public static MetroJaxWsTracer tracer() {
+    return TRACER;
+  }
+
+  public void startSpan(WSEndpoint endpoint, Packet packet) {
+    String serviceName = endpoint.getServiceName().getLocalPart();
+    String operationName = packet.getWSDLOperation().getLocalPart();
+    String spanName = serviceName + "/" + operationName;
+    Context context = startSpan(spanName, INTERNAL);
+    Scope scope = context.makeCurrent();
+
+    // store context and scope
+    packet.addSatellite(new TracingPropertySet(context, scope));
+
+    Span serverSpan = getCurrentServerSpan();
+    if (serverSpan != null) {
+      String serverSpanName = spanName;
+      HttpServletRequest request = (HttpServletRequest) packet.get(MessageContext.SERVLET_REQUEST);
+      if (request != null) {
+        String servletPath = request.getServletPath();
+        if (!servletPath.isEmpty()) {
+          String pathInfo = request.getPathInfo();
+          if (pathInfo != null) {
+            serverSpanName = servletPath + "/" + spanName;
+          } else {
+            // when pathInfo is null then there is a servlet that is mapped to this exact service
+            // servletPath already contains the service name
+            serverSpanName = servletPath + "/" + operationName;
+          }
+        }
+      }
+      serverSpan.updateName(ServletContextPath.prepend(context, serverSpanName));
+    }
+  }
+
+  public void end(Packet packet) {
+    end(packet, null);
+  }
+
+  public void end(Packet packet, Throwable throwable) {
+    Scope scope = (Scope) packet.get(TracingPropertySet.SCOPE_KEY);
```

Comment:
Properties are not removed from the `packet` (unlike other JAX-WS instrumentations), is that okay?

Reply:
For axis and cxf I removed the properties to ensure that we won't end up closing the same span twice in case these interceptors are called twice for whatever reason. On metro these properties can not be removed, remove method exists, but calling it will complain that the property is read only. I guess if needed I could add a property with mutable value that could track whether end has already been called.